llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -614,6 +619,135 @@ static string_vector saveResultSymbolsLists(string_vector &ResSymbolsLists) {
+
+  bool SpecConstsMet = false;
+  bool SetSpecConstAtRT = DoSpecConst && (SpecConstLower == SC_USE_RT_VAL);
+
+  if (DoSpecConst) {
+    // perform the spec constant intrinsics transformation and enumeration on
+    // the whole module
+    ModulePassManager RunSpecConst;
+    ModuleAnalysisManager MAM;
+    SpecConstantsPass SCP(SetSpecConstAtRT);
+    // Register required analysis
+    MAM.registerPass([&] { return PassInstrumentationAnalysis(); });
+    RunSpecConst.addPass(SCP);
+    if (!DoSplit)
+      // This pass deletes unreachable globals. Code splitter runs it later.
+      RunSpecConst.addPass(GlobalDCEPass());
+    PreservedAnalyses Res = RunSpecConst.run(*M, MAM);
+    SpecConstsMet = !Res.areAllPreserved();
+  }
+  if (IROutputOnly) {
+    // the result is the transformed input LLVMIR file rather than a file table
+    saveModule(*M, OutputFilename);
+    return TblFiles;
+  }
+  if (DoSplit) {
+    splitModule(*M, GlobalsSet, ResultModules);
+    // post-link always produces a code result, even if it is unmodified input
+    if (ResultModules.size() == 0)
+      ResultModules.push_back(std::move(M));
+  } else
+    ResultModules.push_back(std::move(M));
+
+  {
+    // reuse input module if there were no spec constants and no splitting
+    string_vector Files =
+        SpecConstsMet || (ResultModules.size() > 1) || SyclAndEsimdKernels
+            ? saveResultModules(ResultModules, IsEsimd)
+            : string_vector{InputFilename};
+    // "Code" column is always output
+    std::copy(Files.begin(), Files.end(),
+              std::back_inserter(TblFiles[COL_CODE]));
+  }
+
+  {
+    ImagePropSaveInfo ImgPSInfo = {
+        true,          DoSpecConst,         SetSpecConstAtRT,
+        SpecConstsMet, EmitKernelParamInfo, IsEsimd};
+    string_vector Files = saveDeviceImageProperty(ResultModules, ImgPSInfo);
+    std::copy(Files.begin(), Files.end(),
+              std::back_inserter(TblFiles[COL_PROPS]));
+  }
+  if (DoSymGen) {
+    // extract symbols per each module
+    collectSymbolsLists(GlobalsSet, ResultSymbolsLists);
+    if (ResultSymbolsLists.empty()) {
+      // push empty symbols list for consistency
+      assert(ResultModules.size() == 1);
+      ResultSymbolsLists.push_back("");
+    }
+    string_vector Files = saveResultSymbolsLists(ResultSymbolsLists, IsEsimd);
+    std::copy(Files.begin(), Files.end(),
+              std::back_inserter(TblFiles[COL_SYM]));
+  }
+  return TblFiles;
+}
+
+using ModulePair = std::pair<std::unique_ptr<Module>, std::unique_ptr<Module>>;
+
+// This function splits a module with a mix of SYCL and ESIMD kernels
+// into two separate modules.
+static ModulePair splitSyclEsimd(std::unique_ptr<Module> M) {
+  // Collect information about the SYCL and ESIMD kernels in the module.
+  std::vector<Function *> SyclKernels;
+  std::vector<Function *> EsimdKernels;
+  for (auto &F : M->functions()) {
+    if (F.getCallingConv() == CallingConv::SPIR_KERNEL) {
```

Comment:
I think SYCL_EXTERNAL functions are missing. They are also "roots" of the modules. See also comment about SYCL_EXTERNAL above. We might not be supporting them properly in sycl-post-link at all.

On the other hand, post-link happens to be invoked on already linked code, so this is not high-priority. A TODO + issue would be enough in that case, I think

Reply:
I'm not sure I understand what would be the scenario with SYCL_EXTERNAL functions and what we should do with them.

Reply:
The scenario is when sycl-post-link is run on a static library with SYCL_EXTERNAL functions of both kinds - ESIMD and SYCL.

Reply:
We discussed it and decided to leave it as it is for now.

Reply:
right, seems a very-low priority topic, as post-link is to be invoked on already linked code